### ISSUE TITLE
widget_update_callback not working?

### ISSUE THREAD

**Issue by phpbits** [siteorigin/siteorigin-panels#207]
Hi,

Thank you very much for adding in_widget_form action. But the data is not saving via widget_update_callback filter. Do you have custom filter to save the data from the fields added via in_widget_form action? Thanks!

**Comment by phpbits** [comment]
Seems to be reported a year ago here : https://wordpress.org/support/topic/page-builder-and-widget-css-classes-v2

**Comment by phpbits** [comment]
Hi,

Any news :) Thanks!

**Comment by Misplon** [comment]
Hi @phpbits Thanks for reaching out. The team are flat out on a series of updates at the moment. We'd definitely like to solve this for you but it will, unfortunately, take us a little more time to get to this. Wish we could get to it right away. Sorry I don't have better news.

**Comment by phpbits** [comment]
Hi @Misplon 

I totally understand :) Just let me know once this feature has been added. Thanks!

**Comment by Misplon** [comment]
Thanks for your understanding. Will do. This issue report will be updated when there is movement. Thanks again.

**Comment by phpbits** [comment]
@Misplon just checking if this feature was integrated and will be very happy to check it. Thanks!

**Comment by Misplon** [comment]
@phpbits Unfortunately, not yet. Wish I had more. The team has been super busy getting our new offering SiteOrigin Premium ready for launch. Once that's out next week and any important bugs are resolved then I'm sure the team and Greg will have more time to dive into tasks that have been waiting.

**Comment by phpbits** [comment]
Thank you very much @Misplon and goodluck with your new product :)

**Comment by Misplon** [comment]
Thanks for your support :) Chat soon.

**Comment by phpbits** [comment]
You're welcome :) I just want to integrate this plugin of mine : https://wordpress.org/plugins/widget-options/ perfectly for more widget control & management ;) Thanks!

**Comment by Misplon** [comment]
That would be awesome. Greg is keen to tackle this as soon as we get our core, mission critical stuff shipped. Please, feel free to follow-up with me in a week or two and I'll follow-up with Greg then.

**Comment by phpbits** [comment]
Sounds perfect :) Just let me know how it goes. Thank you very much!

Cheers!

**Comment by phpbits** [comment]
Hi @Misplon ,

Just following up how is it going ;) Thanks!

Cheers,
Jeffrey

**Comment by Misplon** [comment]
Apologies for the delay, I've asked Greg to take a look as soon as possible.

**Comment by phpbits** [comment]
Thanks :) Will check it now.

Cheers!

**Comment by phpbits** [comment]
@Misplon it seems that you are using **number** parameter instead of _id_ under the widget values :

`Tribe_Image_Widget Object ( [id_base] => widget_sp_image [name] => Image Widget [widget_options] => Array ( [classname] => widget_sp_image [customize_selective_refresh] => [description] => Showcase a single image with a Title, URL, and a Description ) [control_options] => Array ( [id_base] => widget_sp_image ) [number] => c60 [id] => temp [updated] => [option_name] => widget_widget_sp_image )`

Are there any condition I can use on in_widget_form action to check if the widget is being opened under the pagebuilder? Thanks!

**Comment by gregpriday** [comment]
As far as I remember, the `number` attribute in the standard WordPress interface is just a unique index. Instead, Page Builder uses a backbone model ID. I haven't had the chance to double check, but I think checking for a `c` character at the start of the `number` would be a fairly reliable check of whether the widget is coming from Page Builder.

I'll see if we can add another attribute to the widget object that more clearly defines it as coming from Page Builder. As for `id` being `temp`, I'm not sure why that's the case. I'll look into it.

**Comment by phpbits** [comment]
Thanks @gregpriday, I've tried `widget_update_callback` but it's not working on the Page Builder since it's using page id on the widget fields name, even changing the `temp` to correct `id` is not saving it.

**Comment by gregpriday** [comment]
I'll have to investigate how to get it all working properly. It might take some changes to how Page Builder handles widget saving.

As @Misplon mentioned, we're busy working on launching a new product which will allow us to dedicate a lot more resources to improving Page Builder. As soon as we've done that, I'll take a proper look at getting this filter working properly.

**Comment by gregpriday** [comment]
In the mean time, can you send me a copy of your widget or a link to your GitHub repo. It'll be useful for me to have something to experiment with.

**Comment by Misplon** [comment]
@gregpriday Here we go: https://github.com/phpbits/widget-options.

@phpbits To explain a bit further, my estimated timelines for when SiteOrigin Premium would launch and when Greg would have more time resources to look into this were quite far off. The commit above was a quick attempt to get this working but it looks like it'll take a lot more doing. We'll revisit this as soon as we're able but unfortunately don't have any timeline on this at the moment. Sorry, wish we had more right now.

**Comment by phpbits** [comment]
Hi,

Thank you very much for detailed response, just let me know when you've already figured it out. For the meantime I'm not showing the Widget Options when the widget is inside the Page Builder. Would be a great addition for the users to have this options though and I'm looking forward to integrate it perfectly on your plugin. Thanks again!

Cheers,
Jeffrey

**Comment by phpbits** [comment]
Hi,

Would like to follow up this one. Thanks!

Cheers,
Jeffrey

**Comment by braamgenis** [comment]
Hi @phpbits 

I've spent a bit of time trying to figure out why your form values aren't saving and as far as I can tell, it's because they're not compatible with how Page Builder collects and saves form values.

You can see how Page Builder does this [here](https://github.com/siteorigin/siteorigin-panels/blob/2.4.17/js/siteorigin-panels.js#L4780). If you inspect the `data` variable just before it's returned from that function, you'll find that `extended_widget_opts` is under a property called `temp` which is your placeholder id, I think. If you make sure that `temp` ends up under the `widgets` property instead, things should work as expected.

Hope that helps.

**Comment by phpbits** [comment]
Hi @braamgenis 

I'm saving the values on widget_update_callback filter, are you calling it on save_post where you are saving the pagebuilder widgets data? I can't seem to find it. Probably the reason why the values aren't saving. Thanks!

Cheers,
Jeffrey

**Comment by phpbits** [comment]
@braamgenis you can also check this reported issue : https://github.com/siteorigin/siteorigin-panels/issues/251 . Thanks!

**Comment by braamgenis** [comment]
@phpbits We call `widget_update_callback` in `siteorigin_panels_process_raw_widgets` [here](https://github.com/siteorigin/siteorigin-panels/blob/2.4.17/siteorigin-panels.php#L658) which is in turn called in our `save_post` handler [here](https://github.com/siteorigin/siteorigin-panels/blob/2.4.17/siteorigin-panels.php#L579).

But the problem isn't on the back end, it's on the front end. Your form values never reach the back end because Page Builder's JavaScript doesn't find them in the right place. Check the JavaScript function `getFormValues` which I linked to in my previous post, to see how Page Builder collects form values. In future we'd like to make that function smarter to avoid issues like the one you're having, but unfortunately for now you'll have to make sure your form inputs are compatible.

**Comment by phpbits** [comment]
@braamgenis will making the field name something like this : widgets[c66][extended_widget_opts][clearfix][desktop] works? I'll change my field names when inside the builder but if this won't work, I'll just wait for the compatibility to make my plugin works perfectly on the page builder. Thanks!

Cheers,
Jeffrey

**Comment by phpbits** [comment]
@braamgenis I think I'm getting some ideas on the workaround, question though; on in_widget_form action do you have a function to determine if it was loaded on the pagebuilder? Thanks!

**Comment by phpbits** [comment]
@braamgenis I think I've found a workaround to save the field values :) Now I just realized that you don't have widget_display_callback filter when calling the widgets on the frontend. Do you have specific function I can use? Thanks!

**Comment by AlexGStapleton** [comment]
Hi phpbits,

Very sorry for the delay!

Do you still need assistance with this issue or were you able to find a workaround?
To answer your specific question regardless, that's definitally possible. [Here's the code](https://github.com/siteorigin/siteorigin-panels/blob/2.6.5/inc/renderer.php#L431) we use to render widgets. We've added a number of differnet filters that you should be able to use to do what your after.

For example, you can use  [siteorigin_panels_widget_object](https://github.com/siteorigin/siteorigin-panels/blob/2.6.5/inc/renderer.php#L441) to alter the general settings of the widget itself and the specific widget instance.

You could also use [siteorigin_panels_the_widget_html](https://github.com/siteorigin/siteorigin-panels/blob/2.6.5/inc/renderer.php#L515) to completely change how SiteOrigin Page Builder renders your widget.

**Comment by scottsawyer** [comment]
@phpbits I am having a similar issue ( at least I think it's similar ) with a plugin I created and I am totally interested in your workaround.  

The problem I am having is PB essentially prefixes / changes my field names, so when my ajax function which adds more settings, which populates the wrong field names for the current widget., which causes the new fields to not save.

This is from MY_Widget->form()
```
<div class="field-group" id="<?= $this->id . '-' . $group_id; ?>">
                  <input type="hidden" name="<?= $this->get_field_name( 'group' ); ?>[<?= $group_id; ?>]" value="<?= $group_id; ?>" data-group="<?= $group_id; ?>" data-widget="<?= $this->id; ?>">
```
This snippet is what I see when I add my widget to PB and edit for the first time:
```
   <div class="field-group" id="temp-0">
                  <input type="hidden" name="widgets[c42][group][0]" value="0" data-group="0" data-widget="temp">
```
This is what I see when I drop the same widget in a normal sidebar:
```
<div class="field-group" id="wp-builder-customizer-widgets-2-0">
                  <input type="hidden" name="widget-wp-builder-customizer-widgets[2][group][0]" value="0" data-group="0" data-widget="wp-builder-customizer-widgets-2">
```

When I click "Add Fields", this is the relevant part of my ajax handler to add more option groups:
```
// Ajax add fields
    $data = json_decode( stripcslashes( $_POST['data'] ) );

    $widget_id = $data->widget_id;
    $group_id = $data->group_id;

    ++$group_id;
    
    $widget_instance_parts = explode( '-', $widget_id );
    
    $instance_id = array_pop( $widget_instance_parts );
    $instance = 'widget-' . implode( '-', $widget_instance_parts );

    ?>
      <div class="field-group" id="<?= $widget_id . '-' . $group_id; ?>">
        <input type="hidden" name="<?= $instance; ?>[<?= $instance_id; ?>][group][]" value="<?= $group_id; ?>" data-group="<?= $group_id; ?>" data-widget="<?= $widget_id; ?>">
  // .....
```
This is the markup I am getting something like this back from my Add Fields in a normal sidebar
```
<div class="field-group" id="wp-builder-customizer-widgets-2-1">
        <input type="hidden" name="widget-wp-builder-customizer-widgets[2][group][]" value="1" data-group="1" data-widget="wp-builder-customizer-widgets-2">
```
This is what I get in PB:
```
<div class="field-group" id="temp-1">
        <input type="hidden" name="widget-[temp][group][]" value="1" data-group="1" data-widget="temp">
```

Basically, I think I need to get the current field names assigned by PB and use those instead of the field names I am generating.

**Comment by scottsawyer** [comment]
Never mind, I figured out that I should just send back the entire field name via ajax, then parse the field name to build my new name.  It's ugly, but it's the best I can do at the moment:

```
   $prefix = $data->prefix;

    parse_str( $prefix, $prefix_parts);
 
    $prefix_parts = $this->array_keys_multi( $prefix_parts );  // ref: https://gist.github.com/JohnQUnknown/8761761
    array_pop( $prefix_parts );

    $instance = $prefix_parts[0] . '[' . $prefix_parts[1] . ']';
 ?>
      <div class="field-group" id="<?= $widget_id . '-' . $group_id; ?>">
        <input type="hidden" name="<?= $instance; ?>[group][]" value="<?= $group_id; ?>" data-group="<?= $group_id; ?>" data-widget="<?= $widget_id; ?>">
```
That is a lot of gunk, I think I am missing something that would make this easier.  But, now my values are saving, so it's a win.
